Step2 string calcurator

DevVenusK opened the pull request:
## 생각거리
- 하나의 메소드를 작성한 뒤 Commit 하는 걸 잊지말자
- 소스트리는 사용하기 어렵구나.. (revert 죄송합니다.. (__)
- Protocol 을 어떻게 나눠(?) 작성할 것인가.
- Naming..Naming..Naming..Naming.........................
- BigO 에 대한 고민..
- Array 애서 removeFirst() 와 index 로 접근하는 것 중 성능이 빠른것은?

DevVenusK committed:
Implementation of the test teaser program

DevVenusK committed:
Implementation of the test teaser program

DevVenusK committed:
Revert "Implementation of the test teaser program"

This reverts commit 0e0335b966dc76167d1e64523bed9e4afdff0a5e.

DevVenusK committed:
Revert "Implementation of the test teaser program"

This reverts commit 0e0335b966dc76167d1e64523bed9e4afdff0a5e.

DevVenusK committed:
Revert "Implementation of the test teaser program"

This reverts commit 0e0335b966dc76167d1e64523bed9e4afdff0a5e.

DevVenusK committed:
Revert "Implementation of the test teaser program"

This reverts commit 0e0335b966dc76167d1e64523bed9e4afdff0a5e.

DevVenusK committed:
Revert "Implementation of the test teaser program"

This reverts commit 0e0335b966dc76167d1e64523bed9e4afdff0a5e.

DevVenusK committed:
Revert "Implementation of the test teaser program"

This reverts commit 0e0335b966dc76167d1e64523bed9e4afdff0a5e.

DevVenusK committed:
Revert "Revert "Implementation of the test teaser program""

This reverts commit 7910dd1c968fd0fcb7c466ecfcf3fd7628a03eb9.

DevVenusK committed:
Revert "Revert "Implementation of the test teaser program""

This reverts commit 7910dd1c968fd0fcb7c466ecfcf3fd7628a03eb9.

DevVenusK committed:
Revert "Revert "Implementation of the test teaser program""

This reverts commit 7910dd1c968fd0fcb7c466ecfcf3fd7628a03eb9.

DevVenusK committed:
Revert "Revert "Implementation of the test teaser program""

This reverts commit 7910dd1c968fd0fcb7c466ecfcf3fd7628a03eb9.

DevVenusK committed:
Revert "Revert "Implementation of the test teaser program""

This reverts commit 7910dd1c968fd0fcb7c466ecfcf3fd7628a03eb9.

DevVenusK committed:
Revert "Revert "Implementation of the test teaser program""

This reverts commit 7910dd1c968fd0fcb7c466ecfcf3fd7628a03eb9.

DevVenusK committed:
Merge branch 'devvenusk' of https://github.com/DevVenusK/swift-racingcar into devvenusk

DevVenusK committed:
Merge branch 'devvenusk' of https://github.com/DevVenusK/swift-racingcar into devvenusk

DevVenusK committed:
Merge branch 'devvenusk' of https://github.com/DevVenusK/swift-racingcar into devvenusk

DevVenusK committed:
Merge branch 'devvenusk' of https://github.com/DevVenusK/swift-racingcar into devvenusk

DevVenusK committed:
Merge branch 'devvenusk' of https://github.com/DevVenusK/swift-racingcar into devvenusk

DevVenusK committed:
feat(project): StringCalculator 파일 생성

DevVenusK committed:
feat(protocol): StringCalculator Protocol 생성

DevVenusK committed:
feat(method): 사칙연산 메소드 구현

DevVenusK committed:
feat(method): StringCaluratorProtocol 구현

DevVenusK committed:
AruthmeticProtocol 생성

DevVenusK committed:
feat: Error 타입 추가

DevVenusK committed:
feat: Calcurate 함수 리턴타입 추가 및 구현

DevVenusK committed:
Code Clean up

DevVenusK committed:
Add File

DevVenusK committed:
feat: TC 작성

wonhee009 commented on RacingCar/StringCalculator.swift:
```diff
@@ -0,0 +1,106 @@
+//  Created by 김효성 on 2021/10/28.
+//
+
+/*
+ 사용자가 입력한 문자열 값에 따라 사칙연산을 수행할 수 있는 계산기를 구현해야 합니다.
+ 입력 문자열의 숫자와 사칙 연산자 사이에는 반드시 빈 공백 문자열이 있다고 가정합니다.
+ 나눗셈의 경우 결과 값을 정수로 떨어지는 값으로 한정합니다.
+ 문자열 계산기는 사칙연산의 계산 우선순위가 아닌 입력 값에 따라 계산 순서가 결정됩니다.
+ 즉, 수학에서는 곱셈, 나눗셈이 덧셈, 뺄셈 보다 먼저 계산해야 하지만 이를 무시합니다.
+ 예를 들어 2 + 3 * 4 / 2와 같은 문자열을 입력할 경우 2 + 3 * 4 / 2 실행 결과인 10을 출력해야 합니다.
+ */
```
프로젝트 요구사항인거 같은데요.
해당 주석이 반드시 필요한걸까요?🥲

DevVenusK replied:
제거했습니다!

wonhee009 commented on RacingCar/Calculator.swift:
```diff
@@ -13,12 +13,4 @@ struct Calculator {
     func substract(_ lhs: Int, _ rhs: Int) -> Int {
         return lhs - rhs
     }
-    
-    func multiply(_ lhs: Int, _ rhs: Int) -> Int {
-        return lhs * rhs
-    }
-    
-    func devide(_ lhs: Int, _ rhs: Int) -> Int {
-        return lhs / rhs
-    }
```
`Calculator.swift`는 사용하는 곳이 없는거 같은데 프로젝트에서 제거해도 되지 않을까요?

DevVenusK replied:
Step 1 을 진행하던 코드인 것 같아요. 소스트리에서 이것저것 만지다가 꼬인 것 같습니다. 이부분은 제거해둘게요 !

wonhee009 commented on RacingCar/StringCalculator.swift:
```diff
@@ -0,0 +1,106 @@
+        case mutiply = "X"
+        case devide = "/"
+        
+        static func factory(_ string: String) -> ArithmeticOperations? {
+            let arithmeticOperation  = ArithmeticOperations
+                .allCases
+                .filter { $0.rawValue == string }
+                .first
+            return arithmeticOperation
+        }
```
`factory()`에서는 입력 받은 String에 해당하는 Operation case를 반환만 해주는거로 보이는데요.
```swift
Operation(rawValue: "+")
```
를 사용해보시는건 어떨까요?
그러면 Operation이 CaseIterable을 채택할 필요도 없을거 같고, `factory()` 없이 사용 가능할거 같습니다.

wonhee009 commented on RacingCar/StringCalculator.swift:
```diff
@@ -0,0 +1,106 @@
+    
+    func calcurate(_ string: String?) throws -> Int {
+        guard let string = string else { throw InputError.nil }
+        guard !string.isEmpty else { throw InputError.empty }
```
```swift
extension String {
     var isNotEmpty: Bool
}
```
을 정의해서 사용하는 것도 가독성을 높일 수 있는 하나의 방법일 수 있을거 같아요!

wonhee009 commented on RacingCar/StringCalculator.swift:
```diff
@@ -0,0 +1,106 @@
+        
+        guard !numbers.isEmpty else { throw InputError.noNumbers }
+        
+        let operators = string
+            .split(separator: " ")
+            .filter { $0 == "+"
+                || $0 == "-"
+                || $0 == "X"
+                || $0 == "/" }
+            .map { String($0) }
+            .map { ArithmeticOperations.factory($0) }
```
<img width="765" alt="image" src="https://user-images.githubusercontent.com/30456206/139657286-fae49420-33a0-422f-acf8-5c6af89e3377.png">
저는 여기서 타입 추론이 안되고 있다는데 확인 한 번 부탁드려요!🙏

DevVenusK replied:
이부분은 이해가 잘 되질 않네요. 아무리 해봐도 잘되는데.. 조금 더 살펴봐야할 것 같습니다,.

wonhee009 commented on RacingCar/StringCalculator.swift:
```diff
@@ -0,0 +1,106 @@
+        guard let string = string else { throw InputError.nil }
+        guard !string.isEmpty else { throw InputError.empty }
+        
+        var numbers = string
+            .split(separator: " ")
+            .compactMap { Int($0) }
+        
+        guard !numbers.isEmpty else { throw InputError.noNumbers }
+        
+        let operators = string
+            .split(separator: " ")
+            .filter { $0 == "+"
+                || $0 == "-"
+                || $0 == "X"
+                || $0 == "/" }
+            .map { String($0) }
+            .map { ArithmeticOperations.factory($0) }
+        
+        guard !operators.isEmpty else { throw InputError.noOperators }
```
입력받은 문자열에서 숫자와 연산자를 추출하는 과정으로 보이는데요.
`var numbers = ...`, `let operators = ...`에서 `.split()`가 반복되고 있는거 같습니다.
```swift
let a = string.components(" ")
let numbers = a.compactMap { Int($0) }
```
위처럼 입력받은 문자열을 공백을 기준으로 나눈 배열을 들고서 거기서 숫자와 연산자를 빼올 수 있지 않을까요?

DevVenusK replied:
네 맞습니다. 중복된 친구를 확인하지 못했네요. 감사합니다.

wonhee009 commented on RacingCar/StringCalculator.swift:
```diff
@@ -0,0 +1,106 @@
+        
+        guard !operators.isEmpty else { throw InputError.noOperators }
+        
+        var result = numbers.removeFirst()
```
이 지점이 
> Array 애서 removeFirst() 와 index 로 접근하는 것 중 성능이 빠른것은?

해당 고민 지점이라고 생각되는데요.
[공식 문서](https://developer.apple.com/documentation/swift/array/2886730-removefirst)를 확인해 보면 `removeFirst(_:)`의 복잡도는 O(n)이라고 합니다!

DevVenusK replied:
네 맞습니다. removeFirst 의 복잡도는 O(n) 인데, 이 메소드를 forEach 내부에서 실행하고 있다는걸 깜빡했네요.
index 로 수정하였습니다. 감사합니다.

wonhee009 commented on RacingCar/StringCalculator.swift:
```diff
@@ -0,0 +1,106 @@
+        }
+    }
+    
+    func calcurate(_ string: String?) throws -> Int {
```
`calculate(with string: _)` 메서드에서 하는 일이
- input validation
- component로 input 숫자, 연산자 구분
- 전체 계산

으로 보이는데요.
하나의 메서드가 많은걸 하고 있는거 같아요.
좀 더 역할을 나눌 수 있을까요?

DevVenusK replied:
calculator 라는 하나의 커다란 메서드 안에 
var numbers = organizedArray
            .compactMap { Int($0) }

let operators = organizedArray
            .filter { $0 == "+"
                || $0 == "-"
                || $0 == "X"
                || $0 == "/" }
            .map { String($0) }
            .map { ArithmeticOperations.factory($0) }

try? operators
            .enumerated()
            .forEach {
                switch $0.element {
                case .add: result = add(result, numbers[$0.offset])
                case .substract: result = substract(result, numbers[$0.offset])
                case .mutiply: result = multiply(result, numbers[$0.offset])
                case .devide: result = devide(result, numbers[$0.offset])
                case .none: throw InputError.noOperators
                }
            }

세가지 작은 메소드로 나눴다고 생각했습니다.
이 부분을 func 으로 메서드를 만들어서 해야하는걸까요 ??

wonhee009 replied:
하나의 메서드는 하나의 역할을 갖는다고 생각해요.
`calculate()`가 갖고 있는 역할들을 다른 메서드로 쪼개서 사용하는건 어떨까요?

wonhee009 commented on RacingCar/StringCalculator.swift:
```diff
@@ -0,0 +1,106 @@
+}
+
+// MARK: - Private Function
+extension StringCalcurator {
+    func add(_ lhs: Int, _ rhs: Int) -> Int {
+        return lhs + rhs
+    }
+    
+    func substract(_ lhs: Int, _ rhs: Int) -> Int {
+        return lhs - rhs
+    }
+    
+    func multiply(_ lhs: Int, _ rhs: Int) -> Int {
+        return lhs * rhs
+    }
+    
+    func devide(_ lhs: Int, _ rhs: Int) -> Int {
+        return lhs / rhs
+    }
+}
```
해당 메서드의 내용은 `StringCalculator`보다는 `ArithmeticProtocol`과 관련이 더 있어 보이는데요.
어떻게 생각하시나요?🤔

DevVenusK replied:
해당 메서드들은 ArithmeticProtocol 에 정의해 놓은 메소드입니다.
ArithmeticProtocol 프로토콜은 클래스에서 채택하고 있어서 조금 헷갈릴 법 하네요.

Extension 에서 채택하도록 수정했습니다.

wonhee009 commented on RacingCar/StringCalculator.swift:
```diff
@@ -0,0 +1,106 @@
+    func calcurate(_ string: String?) throws -> Int
+}
+
+final class StringCalcurator: StringCalcuratorProtocol, ArithmeticProtocol {
```
계산기에 오타가 있는거 같습니다...!
```suggestion
final class StringCalculator: StringCalculatorProtocol, ArithmeticProtocol {
```
(전체적으로 수정 부탁드립니다!)

DevVenusK replied:
수정했습니다.

wonhee009 commented on RacingCarTest/RacingCarTest.swift:
```diff
@@ -20,37 +20,4 @@ class RacingCarTest: XCTestCase {
         XCTAssertGreaterThan(calculator.substract(-2, -5), 0, "-2 - (-5) > 0 실패")
     }
-    
-    func test_multiply_isSuccess() {
-        let calculator: Calculator = Calculator()
-        let firstNumber: Int = 3
-        let secondNumber: Int = 4
-        
-        XCTAssertTrue(calculator.multiply(firstNumber, secondNumber) == 12)
-    }
-    
-    func test_multiply_isFailure() {
-        let calculator: Calculator = Calculator()
-        let firstNumber: Int = 3
-        let secondNumber: Int = 4
-        
-        XCTAssertFalse(calculator.multiply(firstNumber, secondNumber) == 13)
-    }
-    
-    func test_devide_isSuccess() {
-        let calculator: Calculator = Calculator()
-        let firstNumber: Int = 4
-        let secondNumber: Int = 2
-        
-        XCTAssertTrue(calculator.devide(firstNumber, secondNumber) == 2)
-    }
-    
-    func test_devide_isFailure() {
-        let calculator: Calculator = Calculator()
-        let firstNumber: Int = 3
-        let secondNumber: Int = 4
-        
-        
-        XCTAssertFalse(calculator.devide(firstNumber, secondNumber) == 2)
-    }
```
해당 파일에서 메서드들을 삭제하신거 같은데요. 현재 남아 있는 다른 메서드들도 `RacingCarTest`와는 관계가 없어보이는데요!
메서드 위치를 바꾸시거나 삭제해주시면 좋을거 같습니다.

wonhee009 commented on RacingCarTest/StringCalcuratorTest.swift:
```diff
@@ -0,0 +1,93 @@
+        }
+    }
+    
+    func test_inputError_empty() {
+        let stringArithmeticOperation: String? = ""
+        let result = StringCalcurator.InputError.empty
+        
+        do {
+            let _ = try stringCalurator.calcurate(stringArithmeticOperation)
+        } catch {
+            XCTAssertEqual(error as? StringCalcurator.InputError, result)
+        }
+    }
```
빈 문자열 뿐만 아니라 `" "` 이런 경우도 테스트해볼 수 있지 않을까요?

DevVenusK replied:
테스트 진행했습니다!

wonhee009 reviewed (changes requested):
@DevVenusK 수고하셨습니다!
전체적으로 많은 고민이 느껴지는 코드였습니다!

removeFirst나 contain 같은 경우는 코드에 리뷰를 남겨두었습니다:)
mocking을 통해서 테스트하는 것도 좋은 방법이라고 생각합니다. (그런데 InputError가 StringCalculator 내부 Enum으로 있는게 적절한지도 생각해보면 좋을거 같습니다)

DevVenusK committed:
remove: 필요없는 주석 제거

DevVenusK committed:
fix: split(separator: _) 중복 코드 제거

DevVenusK committed:
fix: removeFirst() 대신 index 를 사용하여 복잡도 감소

DevVenusK committed:
test: " " 일경우 case 추가

DevVenusK committed:
fix: ArithmeticProtocol 의 채택 위치를 extension 으로 변경함

DevVenusK committed:
fix: 오타 수정 StringCalcurator -> StringCalculator

DevVenusK committed:
fix: ArithmeticOperations Enum 의 factory 및 CaseIterable 제거

DevVenusK committed:
fix: ArithmeticProtocol 삭제 및 CalculatorProtocl 추가

각각의 Calculator 를 만들어 줌으로써 Calculator 의 추가를 용이하게 함
StringCalculator에 init 으로 의존성 분리

DevVenusK committed:
fix: func calculrate(_ string:) 의 내부를 작은 함수로 분리

작은 함수들은 Protocol 로 인해 외부에서 사용이 제한되어있는데,
private 를 붙여주는게 올바를까요? 아니면 제어접근자를 제거해도 괜찮을까요?

DevVenusK committed:
fix: numbers 를 변수에서 상수로 변경

DevVenusK committed:
fix: calcurate 함수 코드 정리

wonhee009 reviewed (approved):
@DevVenusK 수고하셨습니다!🎉
다음 스텝에서 만나요!
